### PR TITLE
Backtest result reuse

### PR DESCRIPTION
**Pull request by rokups**
## Summary

Implemented reuse of old backtest results when backtest parameters did not change.

## Quick changelog

- Avoid running a backtest and instead reuse old result if backtest parameters did not change.
- Extra file `backtest-result-*.meta.json` is written with backtest run id (hash derived from config + strategy file).
- New command line parameter `--no-cache` for `backtest` subcommand, opting out from cached result reuse.

## What's new?

When we develop a new version of strategy we usually backtest using `--strategy-list` so we can compare against previous versions. In such case we most likely backtest multiple times using same config and old strategy files do not change. This means we can reuse old backtest result instead of redoing same backtest over and over. This is what was implemented here.

### CODE REVIEW

File: freqtrade/commands/cli_options.py
Comment (xmatthias) [inline freqtrade/commands/cli_options.py]:
```suggestion
        help='Do not reuse cached backtest results.',
```
the "even if it exists" sounds strange when reading it out of context in a help message (think - backtesting for the first time)

File: freqtrade/data/btanalysis.py
Comment (xmatthias) [inline freqtrade/data/btanalysis.py]: I think you should be able to use `glob()` here - something like `dirname.glob('backtest-result.*meta.json)` ... which would simplify this whole segment to just one line

File: freqtrade/data/btanalysis.py
Comment (xmatthias) [inline freqtrade/data/btanalysis.py]: Based on the comment above - i think this should be `break` (stop further processing of the loop)- not continue (continue with next result).

File: freqtrade/misc.py
Comment (xmatthias) [inline freqtrade/misc.py]: I would rather stick to pathlib - which should have this available as `filename.suffix` (obviously assuming the input is Path ...)

os.path can have ... questionable behavior on certain platforms (windows, i'm looking at you) - so i don't trust it for anything related to paths anymore

**Comment by rokups** [comment]
Addressed all comments 👍🏻 btw i wasnt sure if `--no-cache` is a good name, but i could not come up with anything better..

File: freqtrade/data/btanalysis.py
Comment (xmatthias) [inline freqtrade/data/btanalysis.py:118]: The Premise of this looks good - but it doesn't seem to work at the moment

I've just ran 3 backtests in a row (all produce the same hash based on the .meta file) - but it doesns't work and always runs into this.
Some debugging shows that the "loading" filename is generated wrongly.
I think for loading, you'll not need to call `get_backtest_metadata_filename` as the ".glob" already returns the correct filename.

Removing that call does load the file - but crashes further down ...

**Comment by rokups** [comment]
Fixing those 👍🏻 

Also bumped into this:
```
2022-01-07 12:22:34,507 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/home/rk/src/freqtrade/freqtrade/main.py", line 37, in main
    return_code = args['func'](args)
  File "/home/rk/src/freqtrade/freqtrade/commands/optimize_commands.py", line 54, in start_backtesting
    backtesting.start()
  File "/home/rk/src/freqtrade/freqtrade/optimize/backtesting.py", line 740, in start
    show_backtest_results(self.config, self.results)
  File "/home/rk/src/freqtrade/freqtrade/optimize/optimize_reports.py", line 826, in show_backtest_results
    table = text_table_strategy(backtest_stats['strategy_comparison'], stake_currency)
  File "/home/rk/src/freqtrade/freqtrade/optimize/optimize_reports.py", line 662, in text_table_strategy
    drawdown = [f'{t["max_drawdown_account"] * 100:.2f}' for t in strategy_results]
  File "/home/rk/src/freqtrade/freqtrade/optimize/optimize_reports.py", line 662, in <listcomp>
    drawdown = [f'{t["max_drawdown_account"] * 100:.2f}' for t in strategy_results]
KeyError: 'max_drawdown_account'
```

Cached backtest result did not have `max_drawdown_account` entry. This isnt a problem for now, but we may consider adding a backtest format version, then re-running of a backtest could be forced when version changes. For now i will add a comment.

Edit: Actually it is a problem. This exception is thrown by `backtesting-show` command when trying to access old result.

**Comment by rokups** [comment]
I added a test, but it fails. For some reason `mocker.patch('freqtrade.misc', get_strategy_run_id=get_strategy_run_id)` fails to mock that function. Any idea why that would be?

**Comment by xmatthias** [comment]
> I added a test, but it fails. For some reason `mocker.patch('freqtrade.misc', get_strategy_run_id=get_strategy_run_id)` fails to mock that function. Any idea why that would be?

wrong syntax - you're using `mocker.patch.multi` syntax on `mocker.patch` ...

try `mocker.patch('freqtrade.misc.get_strategy_run_id', get_strategy_run_id)` (assuming get_strategy_run_id is a Mock object).

> Edit: Actually it is a problem. This exception is thrown by backtesting-show command when trying to access old result.

I'll fix that in develop tonight ... :+1:

**Comment by rokups** [comment]
Right syntax still fails

File: tests/optimize/test_backtesting.py
Comment (xmatthias) [inline tests/optimize/test_backtesting.py]: this mock is "missing" - which means it doesn't apply correctly.

When mocking a directly imported function (like get_strategy_run_id) - you'll usually have to replace it at the import location (freqtrade.optimize.backtesting - not where it's defined).

That mock is however also not applying correctly - most likely because the test-file is importing backtesting stuff before the mock is actually running (which sometimes requires odd workarounds).

I remember i had this once before which led me to move the test to another file - but i can't remember which test that was - probably one that imports within the test function instead of the top of the file).

File: freqtrade/data/btanalysis.py
Comment (xmatthias) [inline freqtrade/data/btanalysis.py]: I don't think a OrderedDict is necessary.
You're not using any "enhanced" functionality of OrderedDict - and regular dics are now also ordered by default (since 3.7 iirc).

File: tests/optimize/test_backtesting.py
Comment (xmatthias) [inline tests/optimize/test_backtesting.py]: The Failure was actually CAUSED by this (after moving the below mock to the lookup location in `.optimize.backtesting`).

This will mess with the whole backtesting module (everything in that file, most likely) - causing subsequent mocks to somehow not work correctly (not sure how/why) - but fixing this mock fixed the test for no apparent other reason.

Comment (rokups) [reply tests/optimize/test_backtesting.py]: Thats right. Its a leftover from attempt to store metadata in same file. Removed that 👍🏻

**Comment by rokups** [comment]
Actually this needs one more fix before merge: before hashing a config we need to unset strategy list.

File: freqtrade/optimize/optimize_reports.py
Comment (xmatthias) [inline freqtrade/optimize/optimize_reports.py]: It's unclear what you mean with this warning (it refers to a method that does not exist).
Is this because of `get_backtest_metadata_filename` - or `strategy_run_id` (or something else)

Also - can you make an example what would need to be the input for this (whatever function?) to break?

File: freqtrade/optimize/optimize_reports.py
Comment (xmatthias) [inline freqtrade/optimize/optimize_reports.py]: wouldn't a simple dict suffice, conisidering they're ordered by default since python 3.6 or 7? (and you don't even care about the sequence of the keys)

File: freqtrade/optimize/optimize_reports.py
Comment (xmatthias) [inline freqtrade/optimize/optimize_reports.py]: i honestly find this assignment slightly missleading, and dangerous to introduce bugs in the future.

I understand it's just a "convenience" to not have to use `result['metadata']['strategy'] =` below - but for someone not 100% aware that there will only be one metadata dict this might be confusing i think.

File: freqtrade/optimize/backtesting.py
Comment (xmatthias) [inline freqtrade/optimize/backtesting.py:785]: Just for my understanding, maybe ...

assuming i backtested `StrategyA` yesterday, and today i use `--strategy-list StrategyA StrategyB` - (nothing else changed, so `StrategyA` is loaded from the prior result) ... 

do we intentionally write the result of StrategyA to disk again? (so it's now in 2 result files)

(maybe to clarify):
* either we want it to write to file again - in which case writing should happen AFTER sorting below.
* or we don't want to write it again - in which case it should happen before combining results

I assume the prior is the right answer (?)

File: freqtrade/optimize/optimize_reports.py
Comment (rokups) [reply freqtrade/optimize/optimize_reports.py]: Leftover from previous hacky parsing of json - removing 👍🏻

File: freqtrade/optimize/optimize_reports.py
Comment (rokups) [reply freqtrade/optimize/optimize_reports.py]: Seems like i switched it to dict and back again, opss. Yeah order is also not important, it used to be important with a hacky metadata parsing from same json file as a result (that is no longer there).

File: freqtrade/optimize/backtesting.py
Comment (rokups) [reply freqtrade/optimize/backtesting.py:785]: I suspect `backtesting-show` would break if we did not copy old backtest result to new result file. But more importantly copying keeps results consistent in the event where user may delete some but not all old backtest results. If it were the case, in example you described user may lose half of the result even though file of that backtest result is present.

**Comment by samgermain** [comment]
Just from quickly glancing at this, the [weighted interval scheduling](https://imgur.com/a/mVC33Uq) problem rings a bell, I thought I would share this in case it's useful

File: freqtrade/optimize/backtesting.py
Comment (rokups) [inline freqtrade/optimize/backtesting.py]: Looks like dicts need sorting too. So do not merge just yet..

**Comment by rokups** [comment]
For time being im done. Let me know if there is anything else that needs to be done. 👍🏻

File: freqtrade/optimize/backtesting.py
Comment (xmatthias) [inline freqtrade/optimize/backtesting.py]: To me, this check is potentially wrong / erroneous

you should rather use `self.timerange` and evaluate if the end-date is present and not after today.

consider the following "workaround":
`--timerange 20220101-20230101` - it would pass your test here just fine (doesn't end with "-") - However - my end-date is 11 months in the future, so i'll be using incremental data "at will".

File: freqtrade/optimize/backtesting.py
Comment (xmatthias) [inline freqtrade/optimize/backtesting.py:758]: I think your own analysis was quite good - this can be worked around eventually (i'm however not sure how to protect from it).

using `timerange` from above as comparison will not protect aganst this:

* backtest with `--timerange 20210101-20220101` (full prior year) with data ending 2021-11-01.
* download data from November until today
* run the same backtest again `--timerange 20210101-20220101`

By comparing the 2 timerange objects (timerange - self.timerange) - you could protect against the first case (not loading cache) - but you'll not know that you had 2 months missing in the "cached" result (nothing in the config or strategy changed - other than the messages at startup).


TBH - i'm not sure we must protect against this ... we can also add a hint in the documentation about this, warning users to use `--no-cache` when the underlying data changed.

**Review by xmatthias — COMMENTED** [review]
Nice enhancement :+1: 

LGTM :+1:

**Review by xmatthias — APPROVED** [review]

**Comment by rokups** [comment]
I had another idea yesterday. Maybe instead of `--no-cache` we should have had `--cache=[none,day,week,month]`? Then cached results from specified period would be loaded, but no older. This would essentially create an expiring cache and help mitigating issues with data updates invalidating cached results.

**Comment by xmatthias** [comment]
it'd be a possibility for sure - but which of the above should be the default? (Potentially none?)

**Comment by rokups** [comment]
`none` or `day` sound like a good idea. `none` keeps current behavior while `day` is a safest form of caching which saves most time. Cant say i have a preference.

**Comment by xmatthias** [comment]
`day` works fine i think ... although it "bite" me already while trying to change something in backtesting (where i wanted to run the same backtest multiple times in a row) - i think that's a "developer edgecase" we should not design our defaults against 😆

**Pull request merged**
